transport/src/main/java/io/netty5/channel/AbstractChannel.java

```diff
@@ -1966,4 +1978,171 @@ boolean completeFailure(ReadHandleFactory.ReadHandle readHandle, Throwable cause
+            this.writeHandle = writeHandle;
+        }
+
+        void processWriteLoop(ChannelOutboundBuffer outboundBuffer) {
+            this.outboundBuffer = outboundBuffer;
+            try {
+                try {
+                    // We checked before is the outboundBuffer is empty so let's use a do-while loop.
+                    do {
+                        doWriteNow(writeSink);
+                    } while (continueWriting() && !outboundBuffer.isEmpty());
+                } catch (Throwable t) {
+                    handleWriteError(t);
+                } finally {
+                    try {
+                        writeLoopComplete(outboundBuffer.isEmpty());
+                    } catch (Throwable cause) {
+                        // Something is really seriously wrong!
+                        closeWithErrorFromWriteFlushed(cause);
+                    }
+                }
+            } finally {
+                writeHandle.writeComplete();
+
+                // It's important that we call this with notifyLater true so we not get into trouble when flush() is
+                // called again in channelWritabilityChanged(...).
+                updateWritabilityIfNeeded(true, true);
+            }
+        }
+
+        /**
+         * Return the estimated maximum number of bytes that can be written with one gathering write operation.
+         *
+         * @return number of bytes.
+         */
+        public long estimatedMaxBytesPerGatheringWrite() {
+            return writeHandle.estimatedMaxBytesPerGatheringWrite();
+        }
+
+        /**
+         * The number of flushed messages that are ready to be written.
+         *
+         * @return the number of messages.
+         */
+        public int size() {
+            return outboundBuffer.size();
+        }
+
+        /**
+         * Return the first message that should be written.
+         *
+         * @return                          the first message.
+         * @throws IllegalStateException    if called after {@link #complete(long, long, int, boolean)}
+         *                                  or {@link #complete(long, Throwable, boolean)} was called.
+         */
+        public Object first() {
+            return outboundBuffer.current();
+        }
+
+        /**
+         * Call {@link Function#apply(Object)} for each message that is flushed until
+         * {@link Function#apply(Object)} returns {@code false} or there are no more flushed messages.
+         *
+         * @param processor                 the {@link Function} to use.
+         * @throws IllegalStateException    if called after {@link #complete(long, long, int, boolean)}
+         *                                  or {@link #complete(long, Throwable, boolean)} was called.
+         */
+        public void forEach(Function<Object, Boolean> processor) {
+            outboundBuffer.forEachFlushedMessage(processor);
+        }
```

Comment:
@chrisvest I was also considering to do something similar to the `BufferComponentIterator` here and so only have one method

Reply:
Note that the `BufferComponentIterator` trick is only allocation free when there's a single component to iterate. When there are multiple, you need to allocate to link them up.

With the `forEach` we at most allocate the processor object.

Like I noted elsewhere, the `Predicate` interface is a function from object to primitive boolean, so if we use that we don't have to do the `result ? Boolean.TRUE : Boolean.FALSE` dance in the processors or the methods below.